src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTagHelperCompletionService.cs

```diff
@@ -0,0 +1,158 @@
+
+            AddAllowedChildrenCompletions(completionContext, elementCompletions);
+
+            if (elementCompletions.Count > 0)
+            {
+                // If there are allowed children, those are the only possible completions.
+                var emptyResult = ElementCompletionResult.Create(elementCompletions);
+                return emptyResult;
+            }
+
+            elementCompletions = completionContext.ExistingCompletions.ToDictionary(
+                completion => completion,
+                _ => Enumerable.Empty<TagHelperDescriptor>(),
+                StringComparer.OrdinalIgnoreCase);
+
+            var possibleChildDescriptors = _tagHelperFactsService.GetTagHelpersGivenParent(completionContext.DocumentContext, completionContext.ContainingTagName);
+            foreach (var possibleDescriptor in possibleChildDescriptors)
+            {
+                var addRuleCompletions = false;
+                var outputHint = possibleDescriptor.TagOutputHint;
+
+                // Filter out catch-all rules because TagHelpers that target attributes only would light up every child tag otherwise. Force those TagHelpers
+                // to have additional requirements before showing them in the element completion list.
+                var nonCatchAllRules = possibleDescriptor.TagMatchingRules.Where(rule => rule.TagName != TagHelperMatchingConventions.ElementCatchAllName);
+                foreach (var rule in nonCatchAllRules)
+                {
+                    if (elementCompletions.ContainsKey(rule.TagName))
+                    {
+                        addRuleCompletions = true;
+                    }
+                    else if (outputHint != null && elementCompletions.ContainsKey(outputHint))
+                    {
+                        // If the possible descriptors final output tag already exists in our list of completions, we should add every representation
+                        // of that descriptor to the possible element completions.
+                        addRuleCompletions = true;
+                    }
+                    else if (!completionContext.InHTMLSchema(rule.TagName))
+                    {
+                        // If there is an unknown HTML schema tag that doesn't exist in the current completion we should add it. This happens for
+                        // TagHelpers that target non-schema oriented tags.
+                        addRuleCompletions = true;
+                    }
+
+                    if (addRuleCompletions)
+                    {
+                        if (elementCompletions.TryGetValue(rule.TagName, out var existingRuleDescriptors))
+                        {
+                            var combinedDescriptors = new HashSet<TagHelperDescriptor>(existingRuleDescriptors);
+                            combinedDescriptors.Add(possibleDescriptor);
```

Comment:
This is a really expensive way to check `Contains`.

This would be way simpler to build this as `Dictionary<string, List<TagHelperDescriptor>>` or `Dictionary<string, HashSet<TagHelperDescriptor>>`.